seqr/views/apis/anvil_workspace_api.py

```diff
@@ -1,57 +1,130 @@
 """APIs for management of projects related to AnVIL workspaces."""
 
 import logging
+import json
+
 from django.contrib.auth.decorators import login_required
+from django.shortcuts import redirect
 
+from seqr.models import Project, CAN_EDIT
+from seqr.views.utils.json_to_orm_utils import create_model_from_json
 from seqr.views.utils.json_utils import create_json_response
+from seqr.views.utils.file_utils import load_uploaded_file
+from seqr.views.utils.terra_api_utils import add_service_account
+from seqr.views.utils.pedigree_info_utils import parse_pedigree_table
+from seqr.views.apis.individual_api import _add_or_update_individuals_and_families
+from seqr.utils.communication_utils import send_load_data_email
+from seqr.views.utils.permissions_utils import google_auth_required, workspace_has_perm
 from settings import API_LOGIN_REQUIRED_URL
 
 logger = logging.getLogger(__name__)
 
 
-@login_required(login_url=API_LOGIN_REQUIRED_URL)
+def get_project_name(namespace, name):
+    return '{} - {}'.format(namespace, name)
+
+
+@google_auth_required
 def anvil_workspace_page(request, namespace, name):
     """
-    Redirect to the loading data from workspace page or redirect to the project if the project exists.
-
-    Args:
-        request (object): Django request object
-        namespace: The namespace (or the billing account) of the workspace
-        name: The name of the workspace. It also be used as the project name
-    Return:
-        Redirect to a page depending on if the workspace permissions and project existance.
-         Error page if the workspace doesn't exist or the user doesn't have a WRITER permission.
-         Redirect to '/create_project_from_workspace/:workspace_name' if project doesn't exist.
-         Redirect to '/project/:projectGuid'
+    Validate the workspace and project before loading data.
 
+    :param request: Django request object
+    :param namespace: The namespace (or the billing account) of the workspace
+    :param name: The name of the workspace. It also be used as the project name
+    :return Redirect to a page depending on if the workspace permissions or project exists
     """
-    # To be implemented
-    return create_json_response({})
+    if not workspace_has_perm(request.user, CAN_EDIT, namespace, name, can_share=True):
+        message = "Missing required permissions for loading data from workspace {}/{}".format(namespace, name)
+        logger.warning(message)
+        return create_json_response({'error': message}, status=400, reason=message)
```

Comment:
The error message is shown on a blank page currently. How should the error message be displayed to the user?

Reply:
if form submission fails the user should be seeing the form page exactly as it was when they hit submit but with a red error bar at the bottom of it displaying the error and/ or a red warning icon next to the submit button. This should happen by default if you throw a submission error on failure, you should not be writing any custom ui code for displaying errors. Heres an example from a different form
<img width="664" alt="Screen Shot 2021-02-05 at 10 58 10 AM" src="https://user-images.githubusercontent.com/24598672/107057217-36b29180-67a1-11eb-8790-5cc9c2112afa.png">

Reply:
I'm not sure I understand the flow correctly. I think there is no form for the URL `https://seqr_url/anvil_workspace/namespace/name`. When user open the URL, the backend will validate the workspace, the project, and the permissions of the user. If passed the validation, it will redirect to `https://seqr_url/create_project_from_workspace/namespace/name`, then it will have form which can display error message.
Maybe, the `anvil_workspace` URL is unnecessary. We can start with the `create_project_from_workspace` URL. I can do all the validation at the beginning of the API.

Reply:
I misread your first comment and was confused, I thought you were talking about a 400 error being returned from `create_project_from_workspace`. Ignore my answer above.

This is not an API view, so it should never be returning json. I think you should raise `PermissionDenied` here and then django should just handle that for you

Reply:
I understand the usage of the URL now. I changed to raise `PermissionDenied` but there is no useful information telling the user what happened and what needs to do on the return page shown below. Should the view handle the exception and return a more descriptive page?
![image](https://user-images.githubusercontent.com/33550059/107547675-b4a8db00-6b9b-11eb-9ba9-1166722d59ab.png)

Reply:
nope, this is fine. The only way to get to this page is by clicking a link, and the link is only shown to people who will have access. So realistically if you are seeing you this you edited the URL, and then we don't owe you a nice user experience